src/SMCE/BoardView.cpp

```diff
@@ -325,43 +325,123 @@ bool FrameBuffer::read_rgb888(std::span<std::byte> buf) {
     return true;
 }
 
-bool FrameBuffer::write_rgb444(std::span<const std::byte> buf) {
+void convert_rgb444_to_rgb888(std::span<const std::byte> source, std::byte* target) {
+    // for each byte in the input, write two bytes to the buffer
+    for (std::byte from : source) {
+        // duplicate each 4 bits
+        *target++ = (from & (std::byte)0b11110000) | (from >> 4);
+        *target++ = (from << 4) | (from & (std::byte)0b00001111);
+    }
+}
+
+// buf has the RGB444 format
+// framebuffer always has RGB888 format
+bool FrameBuffer::write_rgb444(std::span<const std::byte> source) {
     if (!exists())
         return false;
 
     auto& frame_buf = m_bdat->frame_buffers[m_idx];
-    if (buf.size() != frame_buf.data.size() / 2)
+    if (source.size() != frame_buf.data.size() / 2)
         return false;
 
     [[maybe_unused]] std::lock_guard lk{frame_buf.data_mut};
 
-    auto* to = frame_buf.data.data();
-    for (std::byte from : buf) {
-        *to++ = from & std::byte{0xF};
-        *to++ = from << 4; // Might be a bug there in the case where we have an odd number of pixels in the frame
+    // to is a pointer to the frame buffer data (a long array of bytes, where every 4 bytes is one pixel)
```

Comment:
RGB888 has three 8 bit components, so that's 24 bits per pixel, which equates to 3 bytes, not 4

Reply:
I based my assumption on what you had already written for the RGB444 implementation which converted 1byte into 2bytes (and vice-versa). So if what you're saying now is correct, those implementations never worked either.

Reply:
Bug now tracked here https://github.com/ItJustWorksTM/libSMCE/issues/37; thanks for the report!

Reply:
RGB444 conversions fixed and tested upstream.

Here's the relevant diff if you're interested: https://github.com/ItJustWorksTM/libSMCE/compare/ea5d69c93a2c1b7b510c16eea94f598c5d6f458f...8b643bdd71f16f46b5eb7abc648965eb06cd2634